### PR TITLE
Deprecated synchronous XHR requests rewritten to async

### LINKED ISSUE #8020: Rewrite all synchronous XHR as they are deprecated
Since Chrome 43 I am getting this warning in developer tools:

> Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user's experience. For more help, check http://xhr.spec.whatwg.org/.

I noticed especially in Dashboards plugin some sync requests but there might be even more. We need to rewrite those requests to be async.

I'm tagging it as performance/enhancement since it will also make the UI a bit faster. Feel free to change it.

### PR DESCRIPTION
**Pull request by barbushin**
Fixes #8020

It's the same as https://github.com/piwik/piwik/pull/8205 but excluding changes in `/tests/javascript/index.php` and fixed `widgetMenu.js`.

### CODE REVIEW

File: plugins/Dashboard/javascripts/widgetMenu.js [hunk 1]
```diff
@@ -13,7 +13,10 @@ function widgetsHelper() {
  *
  * @return {object} object containing available widgets
  */
-widgetsHelper.getAvailableWidgets = function () {
+widgetsHelper.getAvailableWidgets = function (callback) {
+    if(!callback) { // @deprecated
+        return this.getAvailableWidgetsSync();
```
Comment (tsteur) [inline plugins/Dashboard/javascripts/widgetMenu.js]: I reckon we don't have to be backwards compatible here, it's not really a public API or so unless some of our code still uses it

File: plugins/Dashboard/javascripts/widgetMenu.js [hunk 2]
```diff
@@ -35,10 +60,33 @@ widgetsHelper.getAvailableWidgets = function () {
...
  *
  * @param {string} uniqueId
+ * @param callback
  * @return {object} widget object
  */
-widgetsHelper.getWidgetObjectFromUniqueId = function (uniqueId) {
-    var widgets = widgetsHelper.getAvailableWidgets();
+widgetsHelper.getWidgetObjectFromUniqueId = function (uniqueId, callback) {
+    if(!callback) { // @deprecated
+        return this.getWidgetObjectFromUniqueIdSync(uniqueId);
```
Comment (tsteur) [inline plugins/Dashboard/javascripts/widgetMenu.js]: Same here and other methods. We can break previous behaviour unless we still use it somewhere ourselves

File: plugins/Morpheus/javascripts/ajaxHelper.js [hunk 3]
```diff
@@ -74,6 +74,7 @@ function ajaxHelper() {
     /**
      * Should ajax request be asynchronous
      * @type {Boolean}
+     * @deprecated Synchronous XHR requests are deprecated(see http://xhr.spec.whatwg.org)
```
Comment (tsteur) [inline plugins/Morpheus/javascripts/ajaxHelper.js]: We should definitely mention this in https://github.com/piwik/piwik/blob/master/CHANGELOG.md under "Deprecations". We should also mention that we will remove it in Piwik 3.0. I created follow up issue https://github.com/piwik/piwik/issues/8246

**Comment by tsteur** [comment]
I tested them seems to work apart from one, nice! 

* Remove dashboard seems to not work anymore (Chrome 43 on Mac)

I have a few comments but they are not necessarily related to this PR, I'm not sure. We could create a follow up issue for this. Same of those issues might have existed before.

* Eg when going to "Admin => General settings" and clicking "Save" (or Sites Management add/edit site or Users Management edit user) it could be nice to change the "state" of the button and use a wording like "Saving" and making the button disabled while the saving is in progress. We recently changed this in the Core Updater as one didn't know whether it is actually doing something or not. It was not really a big issue before as the requests were sync and one couldn't do anything while they were in progress
* When clicking "Save" in "Admin => Plugin settings" it shows "Loading data..." above the "Save" button for a short time. When deleting a user it shows "Loading data..." for a short time as well. Also when pressing "Save" in Personal Settings.

I noticed some more problems but they seem not related to this change: #8245 #8246 #8247

**Comment by tsteur** [comment]
It's nice to have everything async!

**Comment by barbushin** [comment]
@tsteur Thank you for that review! I removed deprecated methods and notices. Also I checked #8245 and #8247 and all of them was reproduced on `master`, so it looks like they are not relative to this Pull request. Anyway, I submited pull requests with fixes for this issues.

**Comment by tsteur** [comment]
Sweet. Did you have a look re removing a dashboard? Does that work for you? 
I will create issues for my other 2 points as they are a problem in general and not really relate to this PR

**Comment by barbushin** [comment]
Do you mean remove widget on Dashboard?
image Looks like it works without any bugs.

**Comment by tsteur** [comment]
Sorry I meant removing an actual dashboard. This can be done in the "Dashboard & Widgets" selector in the bottom left.

**Comment by barbushin** [comment]
I've created new dashboard, and removed, and it works http://i.imgur.com/y9Hs1RF.gif

**Comment by tsteur** [comment]
For some reason it works for me too now. Cannot reproduce it anymore... Good to merge after 2.14.0 release

**Comment by tsteur** [comment]
Actually the Dashboard UI tests are failing, maybe because of this change?

**Comment by mattab** [comment]
What's the status of this PR @barbushin or maybe @tsteur ?

**Comment by barbushin** [comment]
I don't undestand how to see/reproduce UI tests failure. I see links in UI tests results:

> View UI failures (if any) here:
> http://builds-artifacts.piwik.org/ui-tests.master/13864.7/
or: http://builds-artifacts.piwik.org/ui-tests.master/13864.7/screenshot-diffs/diffviewer.html

But it does not works.

**Comment by mattab** [comment]
@barbushin it looks like our screenshot uploader / UI test viewer is buggy for pull requests done in branches outside of Piwik/piwik (it works fine for PR done by core team members). 
can you create a bug report here so we don't forget to fix this issue?

To workaround, could you try to re-create PR on the piwik/piwik branch and then maybe it works then? 

Btw are you able to run UI tests locally and reproduce the UI test failure locally

**Comment by tsteur** [comment]
FYI I created https://github.com/piwik/ui-tests-viewer/issues/11

It should be good to merge but can't see why tests fail. Should ideally only merge when tests are green.

**Comment by barbushin** [comment]
@mattab I re-created PR from `piwik/8020_xhr_sync_fix` branch. There are same UI Tests failing, but it looks like the problem is not in PR code change. Please take a loot at the comments there https://github.com/piwik/piwik/pull/8381#issuecomment-122835011

**Comment by mattab** [comment]
Duplicate of #8381

**Pull request closed without merging**